### PR TITLE
Simple Netgen IO

### PR DESCRIPTION
**Pull request by UZerbinati**

### CODE REVIEW

File: python/dolfinx/io/ngsio.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+# Copyright (C) 2022 Jørgen S. Dokken
```
Comment (jorgensd) [inline python/dolfinx/io/ngsio.py:1]: Change copyright to be under your name;)

File: python/dolfinx/io/ngsio.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+try:
+    import ngsolve as ngs
+    from ngsPETSc.plex import DMPlexMapping
+    import netgen as ng
+    _has_ngs = True
+except ModuleNotFoundError:
+    _has_ngs = False
+
+
+# Map from Gmsh cell type identifier (integer) to DOLFINx cell type
```
Comment (jorgensd) [inline python/dolfinx/io/ngsio.py:33]: Should be ngs cell type?

File: python/dolfinx/io/ngsio.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+# and degree http://gmsh.info//doc/texinfo/gmsh.html#MSH-file-format
+_ngs_to_cells = {(2,3): ("triangle", 1),
+                 (2,4): ("quadrilateral", 1),
+                 (3,4): ("tetrahedron", 1)}
+
+def ufl_mesh(ngs_cell: int, gdim: int) -> ufl.Mesh:
+    """Create a UFL mesh from a Gmsh cell identifier and the geometric dimension.
```
Comment (jorgensd) [inline python/dolfinx/io/ngsio.py:40]: Same comment as above

File: python/dolfinx/io/ngsio.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    print("Shape: {}, Degree: {}".format(shape,degree))
+    cell = ufl.Cell(shape, geometric_dimension=gdim)
+
+    element = basix.ufl.element(
+        basix.ElementFamily.P, cell.cellname(), degree, basix.LagrangeVariant.equispaced, shape=(gdim, ),
+        gdim=gdim)
+    return ufl.Mesh(element)
+
+
+def cell_perm_array(cell_type: CellType, num_nodes: int) -> typing.List[int]:
```
Comment (jorgensd) [inline python/dolfinx/io/ngsio.py:62]: Do we need to expose this function,or should it just be used with `_cpp.perm_gmsh`?

File: python/dolfinx/io/ngsio.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        # We extract topology and geometry
+        if ngmesh.dim == 2:
+            V = ngmesh.Coordinates()
+            T = ngmesh.Elements2D().NumPy()["nodes"]
+            T = np.array([list(np.trim_zeros(a, 'b')) for a in list(T)])-1
+        elif ngmesh.dim == 3:
+            V = ngmesh.Coordinates()
+            T = ngmesh.Elements3D().NumPy()["nodes"]
+            T = np.array([list(np.trim_zeros(a, 'b')) for a in list(T)])-1
+        print("Dimension: {} Points: {}, Num: {}".format(gdim,len(T[0]),T.shape[0]))
```
Comment (jorgensd) [inline python/dolfinx/io/ngsio.py:126]: Remove print statement

File: python/dolfinx/io/ngsio.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            partitioner: Function that computes the parallel
+                distribution of cells across MPI ranks
+
+        Returns:
+            A triplet (mesh, cell_tags, facet_tags) where cell_tags hold
+            markers for the cells and facet tags holds markers for
+            facets (if tags are found in the NetGen model).
+        """
+        # First we generate a mesh
+        ngmesh = geo.GenerateMesh(maxh=hmax)
```
Comment (jorgensd) [inline python/dolfinx/io/ngsio.py:102]: Could you elaborate on how this works with multiple MPI processes? Does it return a distributed mesh?

**Review by jorgensd — CHANGES_REQUESTED** [review]
A few initial comments (I'll have a through look later). Im really excited for this interface!:)

File: python/dolfinx/io/ngsio.py (hunk 2 above)
Comment (UZerbinati) [reply python/dolfinx/io/ngsio.py:33]: Indeed !

File: python/dolfinx/io/ngsio.py (hunk 4 above)
Comment (UZerbinati) [reply python/dolfinx/io/ngsio.py:62]: No need to expose this function you are right. In particular, considering that it's already exposed in the io.gmsh

File: python/dolfinx/io/ngsio.py (hunk 5 above)
Comment (UZerbinati) [reply python/dolfinx/io/ngsio.py:126]: Forgot about that, thank you !

File: python/dolfinx/io/ngsio.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+
+if _has_ngs:
+    __all__ += ["model_to_mesh"]
+
+    def model_to_mesh(geo: ng.libngpy._geom2d, comm: _MPI.Comm, hmax: float, gdim: int = 2,
+                      partitioner: typing.Callable[
+            [_MPI.Comm, int, int, AdjacencyList_int32], AdjacencyList_int32] =
+            create_cell_partitioner(GhostMode.none), transform: typing.Any = None, routine: typing.Any = None) -> typing.Tuple[
+            Mesh, _cpp.mesh.MeshTags_int32, _cpp.mesh.MeshTags_int32]:
```
Comment (IgorBaratta) [inline python/dolfinx/io/ngsio.py:82]: Could you add docstrings for `transform` and `routine`?

**Comment by nate-sime** [comment]
Looks really cool. Is there any way the demos could be merged into one, e.g. [the pyvista demo](https://github.com/FEniCS/dolfinx/blob/main/python/demo/demo_pyvista.py).

**Comment by jorgensd** [comment]
> Looks really cool. Is there any way the demos could be merged into one, e.g. [the pyvista demo](https://github.com/FEniCS/dolfinx/blob/main/python/demo/demo_pyvista.py). 

I think this is a good idea, and would be similar to how we showcase Gmsh

File: python/dolfinx/io/ngsio.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+from dolfinx.mesh import (CellType, GhostMode, Mesh, create_cell_partitioner,
+                          create_mesh, meshtags, meshtags_from_entities)
+
+from mpi4py import MPI as _MPI
+
+__all__ = ["cell_perm_array", "ufl_mesh"]
+
+try:
+    import ngsolve as ngs
+    from ngsPETSc.plex import DMPlexMapping
```
Comment (jorgensd) [inline python/dolfinx/io/ngsio.py:26]: What are the installation intructions for `ngsPETSc`?

File: python/dolfinx/io/ngsio.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+            hmax: The maximum diameter of the elements in the triangulation
+            model: The NetGen model
+            gdim: Geometrical dimension of the mesh
+            partitioner: Function that computes the parallel
+                distribution of cells across MPI ranks
+
+        Returns:
+            A triplet (mesh, cell_tags, facet_tags) where cell_tags hold
+            markers for the cells and facet tags holds markers for
+            facets (if tags are found in the NetGen model).
```
Comment (jorgensd) [inline python/dolfinx/io/ngsio.py:99]: The return type of this function is not correct. The code below does not return cell tags or facet tags.

**Comment by garth-wells** [comment]
@UZerbinati are you planning to keep working on this PR? It would be neat to have the functionality.

**Comment by UZerbinati** [comment]
Dear @garth-wells , yes I plan to keep working on this PR and finish it soon! But after the last NGSolve user meeting we agreed with the NGS dev team to shift most of the NGSolve - PETSc interface (which is used in this PR) to its own repo ([here](https://github.com/NGSolve/ngsPETSc)) and add some functionality (which will be useful for new features in Firedrake), hence this is taking a bit longer. Sorry for the hold-up.

File: python/dolfinx/io/ngsio.py (hunk 8 above)
Comment (UZerbinati) [reply python/dolfinx/io/ngsio.py:26]: https://ngspetsc.readthedocs.io/en/latest/ngsPETSc.html#installation

**Pull request closed without merging**